src/libraries/System.Private.CoreLib/src/System/Number.Dragon4.cs

```diff
@@ -100,203 +101,123 @@ public static unsafe void Dragon4Single(float value, int cutoffNumber, bool isSi
-                    // 3) Set the margin value to the loweset mantissa bit's scale.
-
-                    // scaledValue      = 2 * 2 * mantissa * 2^exponent
-                    BigInteger.SetUInt64(out scaledValue, 4 * mantissa);
-                    scaledValue.ShiftLeft((uint)(exponent));
-
-                    // scale            = 2 * 2 * 1
-                    BigInteger.SetUInt32(out scale, 4);
-
-                    // scaledMarginLow  = 2 * 2^(exponent - 1)
-                    BigInteger.Pow2((uint)(exponent), out scaledMarginLow);
-
-                    // scaledMarginHigh = 2 * 2 * 2^(exponent + 1)
-                    BigInteger.Pow2((uint)(exponent + 1), out optionalMarginHigh);
-                }
-                else                // We have a fractional exponent
-                {
-                    // In order to track the mantissa data as an integer, we store it as is with a large scale
-
-                    // scaledValue      = 2 * 2 * mantissa
-                    BigInteger.SetUInt64(out scaledValue, 4 * mantissa);
-
-                    // scale            = 2 * 2 * 2^(-exponent)
-                    BigInteger.Pow2((uint)(-exponent + 2), out scale);
-
-                    // scaledMarginLow  = 2 * 2^(-1)
-                    BigInteger.SetUInt32(out scaledMarginLow, 1);
-
-                    // scaledMarginHigh = 2 * 2 * 2^(-1)
-                    BigInteger.SetUInt32(out optionalMarginHigh, 2);
-                }
-
-                // The high and low margins are different
-                pScaledMarginHigh = &optionalMarginHigh;
+                mantissaHighBitIdx = DiyFp.DoubleImplicitBitIndex;
+                hasUnequalMargins = (mantissa == (1UL << DiyFp.DoubleImplicitBitIndex));
             }
             else
             {
-                if (exponent > 0)   // We have no fractional component
-                {
-                    // 1) Expand the input value by multiplying out the mantissa and exponent.
-                    //    This represents the input value in its whole number representation.
-                    // 2) Apply an additional scale of 2 such that later comparisons against the margin values are simplified.
-                    // 3) Set the margin value to the lowest mantissa bit's scale.
-
-                    // scaledValue     = 2 * mantissa*2^exponent
-                    BigInteger.SetUInt64(out scaledValue, 2 * mantissa);
-                    scaledValue.ShiftLeft((uint)(exponent));
+                Debug.Assert(mantissa != 0);
+                mantissaHighBitIdx = (uint)BitOperations.Log2(mantissa);
+            }
 
-                    // scale           = 2 * 1
-                    BigInteger.SetUInt32(out scale, 2);
+            Dragon4State state = Dragon4GetScaleValueMargin(mantissa, exponent, mantissaHighBitIdx, hasUnequalMargins, cutoffNumber, isSignificantDigits);
```

Comment:
Here is where I think the logic is most shaky. By grabbing the state of Dragon4 halfway through its run, we have to make a lot of assumptions that the state is exactly how we expect it to be. Based on failing test cases, I think my current understanding of the algorithm is incomplete.

A quick summary of my current understanding: `state.scaledValue` is the value scaled up, and how much it has been scaled is tracked in `state.scale`. `state.digitExponent` tracks the exponent, which I'm currently trying to use to compute the decimal's `scale`.